src/lib/shim/src/tls.rs

```diff
@@ -0,0 +1,610 @@
+    // We need `UnsafeCell` for both so that we can initialize, despite this being
+    // stored in a larger array over which we don't have exclusive access.
+    initd: UnsafeCell<i8>,
+    value: UnsafeCell<MaybeUninit<T>>,
+}
+
+// An initializer for internal use with `LazyLock`. We need an explicit type
+// instead of just a closure so that we can name the type  in `ShimTlsVar`'s
+// definition.
+struct OffsetInitializer {
+    align: usize,
+    size: usize,
+}
+
+impl OffsetInitializer {
+    pub const fn new<T>() -> Self {
+        Self {
+            align: core::mem::align_of::<T>(),
+            size: core::mem::size_of::<T>(),
+        }
+    }
+}
+
+impl lazy_lock::Producer<usize> for OffsetInitializer {
+    // Finds and assigns the next free and suitably aligned offset within
+    // thread-local-storage for a value of type `T`, initialized with function
+    // `F`.
+    fn initialize(self) -> usize {
+        static NEXT_OFFSET: AtomicUsize = AtomicUsize::new(0);
+        let mut next_offset_val = NEXT_OFFSET.load(atomic::Ordering::Relaxed);
+        loop {
+            let overhang = next_offset_val % self.align;
+            let this_var_offset = next_offset_val + self.align - overhang;
+
+            let next_next_offset_val = this_var_offset + self.size;
+            if next_next_offset_val > BYTES_PER_THREAD {
+                panic!("Exceeded hard-coded limit of {BYTES_PER_THREAD} per thread of thread local storage");
+            }
+
+            match NEXT_OFFSET.compare_exchange(
+                next_offset_val,
+                next_next_offset_val,
+                atomic::Ordering::Relaxed,
+                atomic::Ordering::Relaxed,
+            ) {
+                Ok(_) => return this_var_offset,
+                Err(v) => {
+                    // We raced with another thread. This *shouldn't* happen in
+                    // the shadow shim, since only one thread is allowed to run
+                    // at a time, but handle it gracefully. Update the current
+                    // value of the atomic and try again.
+                    next_offset_val = v;
+                }
+            }
+        }
+    }
+}
+
+pub struct ShimTlsVar<T, F = fn() -> T>
+where
+    F: Fn() -> T,
+{
+    // We wrap in a lazy lock to support const initialization of `Self`.
+    offset: LazyLock<usize, OffsetInitializer>,
+    f: F,
+    _phantom: PhantomData<T>,
+}
+// SAFETY: Still `Sync` even if T is `!Sync`, since each thread gets its own
+// instance of the value. `F` must still be `Sync`, though, since that *is*
+// shared across threads.
+unsafe impl<T, F> Sync for ShimTlsVar<T, F> where F: Sync + Fn() -> T {}
+
+impl<T, F> ShimTlsVar<T, F>
+where
+    F: Fn() -> T,
+{
+    /// Create a variable that will be uniquely instantiated for each thread.
+    /// Typically this should go in a `static`.
+    pub const fn new(f: F) -> Self {
+        Self {
+            offset: LazyLock::const_new(OffsetInitializer::new::<T>()),
```

Comment:
Good catch; added more tests to ensure this is caught. It wasn't getting caught before because of some combination of none of the tests allocating more than one thread local variable, that variable type having a large alignment, and the extra space getting wasted in the alignment calculation.